libs/http/include/http/server.hpp

```diff
@@ -95,29 +95,76 @@ class HTTPServer : public AbstractHTTPServer
         acceptor->close(dummy);
       }
     });
+
+    // Since the connection manager contains a reference to this class, we must guarantee it has
+    // destructed before we do.
+    std::shared_ptr<uint64_t> ref_counter = std::make_shared<uint64_t>();
+
+    {
+      auto ref_counter_copy = ref_counter;
+      auto manager          = manager_;
+
+      networkManager_.Post([manager] {
+        auto manager_lock = manager.lock();
+
+        if (manager_lock)
+        {
+          while (manager_lock.use_count() > 1)
+          {
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+          }
+        }
+      });
+    }
+
+    while (ref_counter.use_count() != 1)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    }
   }
 
   void Start(uint16_t port)
   {
-    std::shared_ptr<ConnectionManager> manager   = manager_;
-    std::weak_ptr<Socket> &            socRef    = socket_;
-    std::weak_ptr<Acceptor> &          accepRef  = acceptor_;
-    NetworkManager &                   threadMan = networkManager_;
-
-    networkManager_.Post([&socRef, &accepRef, manager, &threadMan, port] {
-      auto soc   = threadMan.CreateIO<Socket>();
-      auto accep = threadMan.CreateIO<Acceptor>(asio::ip::tcp::endpoint(asio::ip::tcp::v4(), port));
+    std::weak_ptr<ConnectionManager> &manager   = manager_;
+    std::weak_ptr<Socket> &           socRef    = socket_;
+    std::weak_ptr<Acceptor> &         accepRef  = acceptor_;
+    NetworkManager &                  threadMan = networkManager_;
 
-      FETCH_LOG_INFO(LOGGING_NAME,
-                     "Starting HTTPServer on http://127.0.0.1:", accep->local_endpoint().port());
+    // Count instances of this shared pointer that exist to know whether the closure has executed
+    std::shared_ptr<uint64_t> ref_counter = std::make_shared<uint64_t>();
 
-      // allow initiating class to post closes to these
-      socRef   = soc;
-      accepRef = accep;
+    {
+      auto        ref_counter_copy = ref_counter;
+      HTTPServer &server_ref       = *this;
+
+      networkManager_.Post([&socRef, &accepRef, &manager, &threadMan, port, ref_counter_copy,
```

Comment:
Due to these references being passed in, block until there is no possibility the closure can be executed (either io_service dies, was never started, or it is executed)